src/util/png_image.hpp

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include "files.hpp"
+
+#include <png.h>
+
+#include <filesystem>
+#include <new>
+
+namespace image {
+
+enum write_result_t {
+    WR_OK = 0,
+    WR_FOPEN_FAILED,
+    WR_CREATE_WRITE_STRUCT_FAILED,
+    WR_CREATE_INFO_STRUCT_FAILED,
+    WR_SETJMP_PNG_JMPBUF_FAILED,
+    WR_ROW_MALLOC_FAILED,
+};
+
+template <class T>
+write_result_t
+write_image(T image, const std::filesystem::path& path) {
```

Comment:
delete all the commented out `log_e`s you'll do that in the caller with the `write_result_t`

Reply:
Why not just log in this function?

Reply:
Because its a "C-like" function